Support resource attributes in metrics dimensions

albertteoh opened the pull request:
Signed-off-by: Albert Teoh <see.kwang.teoh@gmail.com>

**Description:** 
Adds resource attributes to spanmetrics processor's search space for configured dimension keys, which essentially enables adding "process" (in Jaeger lingo) tags like `region` or `host.name` to metrics labels.

**Link to tracking Issue:** Fixes #4027, and continues from #4052

**Testing:**
- Updated unit tests.
- Integration tested locally and confirmed `ip` and `host.name` labels appear with correct label values when configured in the collector config. Also confirmed default values will be set if the dimension keys don't exist in either span or resource attributes.

**Documentation:** 
- Update README to make it clear that resource attributes are supported.

albertteoh committed:
Support resource attributes in metrics dimensions

Signed-off-by: Albert Teoh <see.kwang.teoh@gmail.com>

mx-psi commented:
@bogdandrutu can you have a look? see https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/4052#issuecomment-886198719

albertteoh committed:
Rebase and fix conflicts

Signed-off-by: Albert Teoh <see.kwang.teoh@gmail.com>

albertteoh commented:
Thanks @mx-psi. @bogdandrutu PTAL; I've rebased and addressed conflicts. The load tests are failing, not sure why, and lack permissions to retry (if it's due to flakiness).

mx-psi commented:
@albertteoh can you fix the merge conflict?

albertteoh committed:
Merge branch 'main' into resource-attr-dims-2

albertteoh commented:
@mx-psi I've addressed the conflicts. It looks like some flaky load tests are failing the build.

albertteoh commented:
Does anyone know why the build is stuck?

mx-psi commented:
Pushing an empty commit or closing and reopening the PR have helped with this in the past (not sure why it's happening)

albertteoh commented:
Thanks @mx-psi, I've tried the close & open approach 🤞🏼

albertteoh commented:
That's looking good now, thanks again @mx-psi. Is anyone able to review please?

albertteoh commented:
Hi @bogdandrutu, PTAL

alolita commented:
@open-telemetry/collector-approvers Can you please review this PR? Thanks.

anuraaga commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -332,19 +332,21 @@ func (p *processorImp) updateLatencyMetrics(key metricKey, latency float64, inde
+		if v, ok := getDimensionValue(d, span.Attributes(), resourceAttrs); ok {
+			dims.Upsert(d.Name, v)
+		} else {
+			p.logger.Debug(fmt.Sprintf("%q metric dimension omitted; not found and no default configured", d.Name),
```
This seems like it could be too spammy even for a debug log

anuraaga commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -360,38 +362,54 @@ func concatDimensionValue(metricKeyBuilder *strings.Builder, value string, prefi
+// an empty string value from a state where no value was found.
+func getDimensionValue(d Dimension, attr pdata.AttributeMap, resourceAttr pdata.AttributeMap) (v pdata.AttributeValue, ok bool) {
+	// The more specific span attribute should take precedence.
+	for _, attrMap := range []pdata.AttributeMap{attr, resourceAttr} {
```
Would generally not use a loop for just checking two items

anuraaga reviewed (approved):
Just some small points looks useful

bogdandrutu commented:
@albertteoh please address feedback

albertteoh committed:
Address review feedback

Signed-off-by: Albert Teoh <see.kwang.teoh@gmail.com>

albertteoh commented:
Thank you @alolita and @anuraaga for the review! 

@bogdandrutu Feedback addressed in 1849592.

albertteoh commented:
Closing and reopening PR to retry failing tailsampling unittest, which may be flaky:

```
race: limit on 8128 simultaneously alive goroutines is exceeded, dying
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor	6.570s
```

tigrannajaryan commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -360,38 +356,55 @@ func concatDimensionValue(metricKeyBuilder *strings.Builder, value string, prefi
-// The metric key is a simple concatenation of dimension values.
-func buildKey(serviceName string, span pdata.Span, optionalDims []Dimension) metricKey {
+// will attempt to add any additional dimensions the user has configured that match the span's attributes
+// or resourceAttr. If the dimension exists in both, the span's attributes, being the most specific, takes precedence.
```
```suggestion
// or resource attributes. If the dimension exists in both, the span's attributes, being the most specific, takes precedence.
```

tigrannajaryan reviewed (approved):
LGTM. Please resolve comments that are already addressed.

albertteoh committed:
Update processor/spanmetricsprocessor/processor.go

Co-authored-by: Tigran Najaryan <4194920+tigrannajaryan@users.noreply.github.com>

albertteoh commented:
Thanks @tigrannajaryan, comments resolved.

albertteoh commented:
> Please file an issue for the flaky test and mark as a bug (include logs, etc to help reproduce).

Issue created: #5678
